pkg/model/components/context.go

```diff
@@ -65,6 +67,20 @@ func KubernetesVersion(clusterSpec *kops.ClusterSpec) (*semver.Version, error) {
 	return sv, nil
 }
 
+// UseEtcdV3 checks to see if etcd v3 is enabled.
+// @question: I kinda feel like this functionality should be on the kops.CluserSpec itself, though I don't
+// know how to do this without duplicating methods and how it effects the versions v1, v2 etc
+func UseEtcdV3(spec *kops.ClusterSpec) bool {
+	// validation ensure the etcd must be both one or the other
```

Comment:
Personally I would probably parse the version, and check if the prefix was `v2.` / `2.` or `v3.` / `3.`.   I suppose there is also the chance that when we mirror the image we will put a prefix on it (e.g. `gcr.io/myimages/etcd_v3.0.17`) but I think we don't do that yet.

Specifying the storage-type separately is sort of appealing here, except that it isn't really the storage-type we're specifying, and as you say we would rather this was on the cluster/apiserver.  The flag is mapped on the apiserver (you set StorageBackend), so we could let the user set it there if they really want to.

I suspect we keep it simple for now, don't add storage-type, parse the version to infer the etcd version using some simple heuristics, and see what is needed from there.

And that implies that we replace `validateEtcdStorage` with a similar loop, but checking that the inferred version of etcd is the same (which also lets us alert the user if they specify a version we don't recognize)